src/cplscheme/BiCouplingScheme.cpp

```diff
@@ -50,9 +50,10 @@ BiCouplingScheme::BiCouplingScheme(
 }
 
 void BiCouplingScheme::addDataToSend(
-    const mesh::PtrData &data,
-    mesh::PtrMesh        mesh,
-    bool                 requiresInitialization)
+    const mesh::PtrData &    data,
+    const time::PtrWaveform &ptrWaveform,
```

Comment:
Why do we have both data and waveform here? Shouldn't the one replace the other?

Reply:
I don't think that this is easily possible. The `Waveform` is currently just a buffer for values to be used for interpolation and extrapolation. The `Data` also stores the id and the name of the data - and of course the `_values` itself.

I guess we could replace `Data::_values` with a pointer to the data stored in the waveform. Then the `Waveform` will become our only store for values. But this is a big step, because `Data::_values` are used and updated almost everywhere (also outside of the time or cplscheme package). I am also not sure whether this is a good idea, because then all packages have to interact with the time package and the `Waveform`.

Currently we are copying `Data::_values` into the `Waveform` on demand. Only the cplscheme packages accesses the waveform. Other packages do not have to worry about the waveform at all, because it is wrapped by the cplscheme. Storing data happens here: 

https://github.com/precice/precice/blob/b784b3849142693c315a16698328bfa975b348e5/src/cplscheme/BaseCouplingScheme.cpp#L265

Reply:
https://github.com/BenjaminRodenberg/precice/pull/4 actually goes some steps into this direction, but entirely removing the `PtrData` is, as explained above, a big step.